griptape/drivers/vector/opensearch_vector_store_driver.py

```diff
@@ -0,0 +1,202 @@
+from griptape import utils
+import logging
+from griptape.drivers import BaseVectorStoreDriver
+from attr import define, field, Factory
+
+
+@define
+class OpenSearchVectorStoreDriver(BaseVectorStoreDriver):
+    host: str = field(kw_only=True)
+    port: int = field(default=443, kw_only=True)
+    http_auth: Optional[Union[str, Tuple[str, str]]] = field(default=None, kw_only=True)
+    use_ssl: bool = field(default=True, kw_only=True)
+    verify_certs: bool = field(default=True, kw_only=True)
+    index_name: str = field(kw_only=True)
+
+    client: OpenSearch = field(default=Factory(
+        lambda self: OpenSearch(
+            hosts=[{'host': self.host, 'port': self.port}],
+            http_auth=self.http_auth,
+            use_ssl=self.use_ssl,
+            verify_certs=self.verify_certs,
+            connection_class=RequestsHttpConnection
+        ),
+        takes_self=True
+    ))
+
+    def upsert_vector(
+            self,
+            vector: list[float],
+            vector_id: Optional[str] = None,
+            namespace: Optional[str] = None,
+            meta: Optional[dict] = None,
+            **kwargs
+    ) -> str:
+
+        vector_id = vector_id if vector_id else utils.str_to_hash(str(vector))
+        doc = {
+            "vector": vector,
+            "namespace": namespace,
+            "metadata": meta
+        }
+        doc.update(kwargs)
+        response = self.client.index(index=self.index_name, id=vector_id, body=doc)
+
+        return response["_id"]
+
+    def load_entry(self, vector_id: str, namespace: Optional[str] = None) -> Optional[BaseVectorStoreDriver.Entry]:
+        try:
+            query = {
+                "bool": {
+                    "must": [
+                        {"term": {"_id": vector_id}}
+                    ]
+                }
+            }
+
+            if namespace:
+                query["bool"]["must"].append({"term": {"namespace": namespace}})
+
+            response = self.client.search(index=self.index_name, body={"query": query, "size": 1})
+
+            if response["hits"]["total"]["value"] > 0:
+                vector_data = response["hits"]["hits"][0]["_source"]
+                entry = BaseVectorStoreDriver.Entry(
+                    id=vector_id,
+                    meta=vector_data.get("metadata"),
+                    vector=vector_data.get("vector"),
+                    namespace=vector_data.get("namespace")
+                )
+                return entry
+            else:
+                return None
+        except Exception as e:
+            logging.error(f"Error while loading entry: {e}")
+            return None
+
+    def load_entries(self, namespace: Optional[str] = None) -> list[BaseVectorStoreDriver.Entry]:
+
+        query_body = {
+            "size": 10000,
+            "query": {
+                "match_all": {}
+            }
+        }
+
+        if namespace:
+            query_body["query"] = {
+                "match": {
```

Comment:
Does having both `match_all` and `match` actually work?

Reply:
Yes!